UnlyEd/next-right-now
# V2 - New preset `v2-mst-aptd-gcms-lcz-sty` with hybrid SSG/SSR instead of SSR-only (based on `v1-ssr-mst-aptd-gcms-lcz-sty`)

# Demos

- https://v2-mst-aptd-gcms-lcz-sty.vercel.app/ uses the latest branch commit
- https://nrn-v2-mst-aptd-gcms-lcz-sty-c1.vercel.app/ is the production environment for Customer 1
- https://nrn-v2-mst-aptd-gcms-lcz-sty-c1-preview.vercel.app/ is the preview/staging environment for Customer 1 (with Preview mode enabled)
- https://nrn-v2-mst-aptd-gcms-lcz-sty-c2.vercel.app/ is the production environment for Customer 2
- https://nrn-v2-mst-aptd-gcms-lcz-sty-c2-preview.vercel.app/ is the preview/staging environment for Customer 2 (with Preview mode enabled)

# Status
This PR/branch is **stable** and has been added as the new official/default NRN preset.

We still have some features to implement, but they're secondary.

# Notes
_This PR is not meant to be merged, but to provide a diff against the base branch (which was a preset using SSR-only, and has been deprecated now that v2 is released)_

PR from #42 (preset branch was renamed, thus forcing us to open a new PR), useful for extra context (discussions about technical implementation)

---

## New features in this PR:
- Language detection and redirect to preferred language (using experimental redirects/rewrites)
- Sub locales support - fr, en, fr-FR and en-US are possible (change url directly to see it in action, or use the footer language icon)
- Use of hooks in favour of HOCs whenever better
- SSG/SSR compatible Apollo client (GraphCMS backend for demo)
- Automatic static optimisation enabled (no _app.getInitialProps)
- SSG-first (my own personal goal is to use 100% SSG web apps, but we do support hybrid apps out of the box)
- Reusable getStaticProps/getServerSideProps (not supported to use them app-wide in _app yet, so we made reusable helpers instead)
- Proper rehydration and server/bowser reconciliation, following https://joshwcomeau.com/react/the-perils-of-rehydration
- Examples with SSG/SSR, and dedicated SSG examples with fallback/revalidate options
- Improved errors handling (top-level, etc.)
- Next.js v9.4.0
- Track perfs statistics "web-vitals-metrics"
- Complete redesign of the demo app (v2)

## Roadmap - In progress


### Roadmap - Backlog
- E2E (tests or redirections, url rewriting, navigation, i18n-related features)


### Done
- ✅ Check TS props types match actual props
- ✅ Tests how errors behave
  - See https://github.com/zeit/next.js/discussions/12562#discussioncomment-12435
- Refactor GQL queries (lots of duplicated code atm)
- ✅ Refactor withApollo `ssr` option which was designed for <v9.3 and doesn't consider `getServerSideProps` option, and is confusing
- ✅ Refactor the `_document`
- ✅ _app using functional component - See https://github.com/UnlyEd/next-right-now/issues/46#issuecomment-627333877
- ✅ Upgrade to Next.js v9.4.0
- ✅ Track perfs statistics "web-vitals-metrics" (new v9.4 feature) https://nextjs.org/blog/next-9-4#integrated-web-vitals-reporting
  - See feedback at https://github.com/zeit/next.js/discussions/12959
- ✅ Improve demo LightHouse performances
  - See https://github.com/UnlyEd/next-right-now/pull/42#issuecomment-629637689
- ✅ 🐛 Support `public` folder properly
  - Fixed by https://github.com/UnlyEd/next-right-now/pull/42/commits/fad1e2df3f56c969a2838ef830a2b6434838e899
- ✅ Upgrade node.js version used by CI (use v12 instead of v10)
- ✅ Change language without full page reload (use router instead)
- ✅ Do not deploy .test functions
- ✅ ❤️ Add more example + restructure examples (UI/UX) - Would love some community help on this, it's basically redesigning all pages and workflow to make it more dev-friendly
  - Status tracked at https://github.com/UnlyEd/next-right-now/issues/45
- ✅ Add `revalidate` example
- ✅ Add `fallback` example
- ✅ Update general documentation about this new preset (v2)
- ✅ New example using **preview mode** for dev/staging environment, will be useful for the SSG page that load data from GraphCMS (e.g: /examples), header/footer components will be pure SSG and won't use preview mode
  - See https://github.com/zeit/next.js/discussions/11552#discussioncomment-12519
  - WIP https://github.com/UnlyEd/next-right-now/pull/70
  - See https://github.com/vercel/next.js/discussions/13589

### Optional other improvements
- Sitemap - See https://github.com/zeit/next.js/discussions/12403 and https://github.com/axross/kohei.dev/blob/master/pages/sitemap.xml.ts
- Reflexion regarding "page dynamic data", to keep them separated from other "common" page data (common to all pages)
  - See https://github.com/zeit/next.js/discussions/12655



### On hold (stuck, any help appreciated)
- Configure CodeSandbox project for easier "playing around" and reproducible example to help 3rd party issues debug
  - WIP, doesn't work properly - See https://github.com/UnlyEd/next-right-now/issues/65
- Optimise bundle (dependencies, e.g: `winston`)
  - Some libs are just too heavy for what benefits they provide, the goal is to only keep what's really useful
  - See https://github.com/UnlyEd/next-right-now/issues/49
  - Opened a feature request https://github.com/vercel/next.js/issues/13594
- "App context" - Clarify combinations of rendering modes and options 
  - The goal is to provide a reliable API (context/hook) to help devs understanding how the page was served and implement different behaviours depending on it 
  - See https://github.com/zeit/next.js/discussions/12558#discussioncomment-12303
  - See https://github.com/zeit/next.js/discussions/12240

### Known bugs
All fixed
